bitcoin/src/address.rs

```diff
@@ -727,34 +820,32 @@ impl Address {
         }
     }
 
-    /// Returns true if the given pubkey is directly related to the address payload.
+    /// Checks whether network of this address is as expected.
     ///
-    /// This is determined by directly comparing the address payload with either the
-    /// hash of the given public key or the segwit redeem hash generated from the
-    /// given key. For taproot addresses, the supplied key is assumed to be tweaked
-    pub fn is_related_to_pubkey(&self, pubkey: &PublicKey) -> bool {
-        let pubkey_hash = pubkey.pubkey_hash();
-        let payload = self.payload.as_bytes();
-        let xonly_pubkey = XOnlyPublicKey::from(pubkey.inner);
-
-        (*pubkey_hash == *payload)
-            || (xonly_pubkey.serialize() == *payload)
-            || (*segwit_redeem_hash(&pubkey_hash) == *payload)
+    /// For details about this mechanism, see section *Parsing addresses* on [`Address'].
+    #[inline]
+    pub fn expect_network(self, expected: Network) -> Result<Address, Error> {
+        if self.is_valid_for_network(expected) {
+            Ok(self.unsafe_mark_network_valid())
+        } else {
+            Err(Error::NetworkValidation { found: self.network, expected })
+        }
     }
 
-    /// Returns true if the supplied xonly public key can be used to derive the address.
+    /// Marks, without any checks, network of this address as valid.
     ///
-    /// This will only work for Taproot addresses. The Public Key is
-    /// assumed to have already been tweaked.
-    pub fn is_related_to_xonly_pubkey(&self, xonly_pubkey: &XOnlyPublicKey) -> bool {
-        let payload = self.payload.as_bytes();
-        payload == xonly_pubkey.serialize()
+    /// This is a life-threatening method, reader will most likely prefer [`expected_network`]
+    /// as a safe variant. For details about this mechanism, see section *Parsing addresses*
+    /// on [`Address'].
+    #[inline]
+    pub fn unsafe_mark_network_valid(self) -> Address {
```

Comment:
The word for this mut not be `unsafe`. `unsafe` has a very specific meaning in Rust and would be confusing. We usually use the word `dangerous`.

Reply:
In this case I think `unchecked` is the idiomatic name.

I'd suggest calling this `to_checked_unchecked` :)

Reply:
Not sure if you're trolling (though it is true `unchecked` is common), I think it should be `to_valid_unchecked` which still looks kinda silly but less than before. Or just `assume_valid`.

Reply:
I'm happy with both of those.

I wasn't trolling, but I also thought that we were using `Checked`/`Unchecked` language rather than `Valid`/`Invalid` language. But it looks like we have a weird mix .. `NetworkValid` and `NetworkUnchecked`.

IMHO we should change to use "checked" and "unchecked" everywhere, since my feeling is that we're checking a specific property of the address rather than checking its validity, but I don't feel strongly at all.

Reply:
Side comment: it is hard to pick a good name for this method until the `NetworkUnchecked`/`NetworkValid` inconsistency is resolved.